Added OTC verification endpoint and sign in redirect params

troyciesco opened the pull request:
closes https://linear.app/ghost/issue/PROD-2497

- adds an endpoint that verifies an OTC and returns a magic link with the token and a hash of the otc. This allows the caller to use the magic link to sign in
- when a user routes to the magic link, the otcVerificationHash is checked - the otc and token are verified, and a timestamp is checked so that the otcVerificationHash is only valid for 5 minutes 
- this pattern allows us to use the same logic for OTC as signing in with a magic link, but with a way to tweak a few things specifically for OTC later, such as handling usage limits, brute force checks, etc.

troyciesco committed:
Added OTC verification endpoint and sign in redirect params

troyciesco committed:
update tests

troyciesco committed:
add util

troyciesco committed:
update tests and comments

troyciesco committed:
improved tests and helpers

troyciesco committed:
more cleanup

troyciesco committed:
PR feedback

troyciesco commented on ghost/core/core/server/services/members/otc-hash-utils.js:
```diff
@@ -0,0 +1,46 @@
+const crypto = require('crypto');
```
def not married to this file name. it also kinda felt more like a helper function than something that belonged on `SingleUseTokenProvider` (and def not `RouterController`) though I could be convinced otherwise

kevinansfield committed:
renamed `getTokenById` to `getTokenByRef`

- matches the methods use-case and hides the `id` as ref implementation detail, reduces churn/confusion when we switch to `uuid` later on

kevinansfield commented on ghost/core/core/server/web/members/app.js:
```diff
@@ -85,6 +85,14 @@ module.exports = function setupMembersApp() {
     );
+    membersApp.post(
+        '/api/verify-otc', 
+        bodyParser.json({limit: '5mb'}),
```
Does this need to be 5mb? I would have thought the [default of 100kb](https://expressjs.com/en/resources/middleware/body-parser.html#limit) would be enough for this endpoint as it only receives a the 6-digit OTC and a reference?

kevinansfield committed:
use MagicLinkService.getSigninURL() in place of duplicated URL building

kevinansfield committed:
fixed otcRef required error message inconsistency

kevinansfield commented on ghost/core/core/server/services/members/members-api/controllers/RouterController.js:
```diff
@@ -649,6 +651,96 @@ module.exports = class RouterController {
+                message: tpl(messages.badRequest),
+                context: 'otc and otc_ref are required'
+            });
+        }
```
We only want to support `otcRef` in the request body. The inconsistent error message has been fixed.

troyciesco committed:
remove limit

kevinansfield commented on ghost/core/core/server/services/members/members-api/controllers/RouterController.js:
```diff
@@ -649,6 +651,85 @@ module.exports = class RouterController {
+            const isValidOTC = await tokenProvider.verifyOTC(otcRef, otc);
+
+            if (!isValidOTC) {
+                res.writeHead(400, {'Content-Type': 'application/json'});
+                return res.end(JSON.stringify({
+                    valid: false,
+                    message: tpl(messages.invalidCode)
+                }));
+            }
+
+            const tokenValue = await tokenProvider.getTokenByRef(otcRef);
+            if (!tokenValue) {
+                res.writeHead(400, {'Content-Type': 'application/json'});
+                return res.end(JSON.stringify({
+                    valid: false,
+                    message: tpl(messages.invalidCode)
+                }));
```
We should be following the newer pattern of `throw new errors.BadRequestError(...` rather than writing to the response directly. The direct `res.*` approach is a very old hangover from when `members-api` was a separate app from Ghost

kevinansfield committed:
updated members-ssr._getMemberDataFromToken typedef

troyciesco commented on ghost/core/core/server/services/members/members-api/controllers/RouterController.js:
```diff
@@ -649,6 +651,85 @@ module.exports = class RouterController {
+        }
+
+        const tokenProvider = this._magicLinkService.tokenProvider;
+        // Validate OTC format using provider (fallback to 6 digits if unavailable)
+        const isFormatValid = tokenProvider?.isOTCFormatValid?.(otc) ?? /^\d{6}$/.test(otc);
+
+        if (!isFormatValid) {
+            throw new errors.BadRequestError({
+                message: tpl(messages.invalidCode)
+            });
+        }
```
i'd gone back and forth on errors, and originally had a specific message if an OTC was too short - but then kind of leaned towards the more generic message (similar to a vague "username or password is incorrect"). 

I was also trying to catch something like this before doing the further verification processing, but it's probably more than necessary.

even besides the fact that the specific length check probably isn't necessary, looking at this now i think it's in the wrong place anyway - verifyOTC in SingleUseTokenProvider would make more sense.

i'm gonna test the error flow again locally (using the otc-portal-submit branch) just to double-check but otherwise will update!

kevinansfield commented on ghost/core/test/unit/server/services/members/members-api/controllers/RouterController.test.js:
```diff
@@ -1097,4 +1097,313 @@ describe('RouterController', function () {
+                mockSettingsCache.get.withArgs('members_email_auth_secret').returns(OTC_TEST_CONSTANTS.VALID_SECRET);
+            });
+
+            it('should return success response with valid inputs', async function () {
+                // use a fake time so we get a stable OTC verification hash for our mocked token/secret
+                sinon.useFakeTimers(new Date('2021-01-01'));
+                const TEST_OTC_VERIFICATION_HASH = '1609459200:c3784e7545cd61c87b34b9bd6d7b840c1225679c74fbc04bc07302a7a1c6aed4';
+
```
This is already handled by the `sinon.restore()` call in the top-level `afterEach` hook

kevinansfield commented on ghost/core/test/unit/server/services/members/members-api/controllers/RouterController.test.js:
```diff
@@ -1097,4 +1097,313 @@ describe('RouterController', function () {
+                    assert.fail('Expected BadRequestError to be thrown');
+                } catch (error) {
+                    assert(error instanceof errors.BadRequestError);
+                    assert.equal(error.message, OTC_TEST_CONSTANTS.ERROR_MESSAGES.BAD_REQUEST);
+                    assert.equal(error.context, OTC_TEST_CONSTANTS.ERROR_MESSAGES.OTC_REQUIRED);
```
We shouldn't ever be testing exact error message content unless they contain dynamic data. If we're expecting a specific error of type `x` then the error should have a unique `code` set on it and tested against.

troyciesco committed:
remove unneeded isOTCFormatValid function

troyciesco committed:
update RouterController to use tokenProvider for creating hash

troyciesco committed:
fix tests

troyciesco committed:
clean up success res

troyciesco commented on ghost/core/core/server/services/members/members-api/controllers/RouterController.js:
```diff
@@ -649,6 +652,70 @@ module.exports = class RouterController {
+            });
+        }
+
+        return res.json({redirectUrl});
```
this included `valid: true, success: true, message: 'OTC verification successful'` before, which i think are all unnecessary? at least it didn't seem like any of that information was used as part of building the magic link. obvs could easily add back if i'm wrong though

kevinansfield committed:
minor cleanup

kevinansfield committed:
added e2e-api test for verify-otc endpoint

kevinansfield committed:
added referer/redirect passthrough for verify-otc endpoint

- updates `/verify-otc` to use the same referer/redirect extraction as `/send-magic-link`
- NOTE: Portal will need to be updated to reflect the change

kevinansfield committed:
fixed wrapped function comment location in members/app.js

kevinansfield committed:
fixed lint

troyciesco committed:
remove unused variables
